source/postprocess/dynamic_topography.cc

```diff
@@ -97,160 +257,190 @@ namespace aspect
                     }
                 }
 
               if (face_idx == numbers::invalid_unsigned_int)
                 continue;
 
-              fe_values.reinit (cell);
-
-              // get the various components of the solution, then
-              // evaluate the material properties there
-              fe_values[this->introspection().extractors.temperature]
-              .get_function_values (this->get_solution(), in.temperature);
-              fe_values[this->introspection().extractors.pressure]
-              .get_function_values (this->get_solution(), in.pressure);
-              fe_values[this->introspection().extractors.velocities]
-              .get_function_values (this->get_solution(), in.velocity);
-              fe_values[this->introspection().extractors.velocities]
-              .get_function_symmetric_gradients (this->get_solution(), in.strain_rate);
-              fe_values[this->introspection().extractors.pressure]
-              .get_function_gradients (this->get_solution(), in.pressure_gradient);
-
-              in.position = fe_values.get_quadrature_points();
-
-              for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
-                fe_values[this->introspection().extractors.compositional_fields[c]]
-                .get_function_values(this->get_solution(),
-                                     composition_values[c]);
-              for (unsigned int i=0; i<fe_values.n_quadrature_points; ++i)
-                {
-                  for (unsigned int c=0; c<this->n_compositional_fields(); ++c)
-                    in.composition[i][c] = composition_values[c][i];
-                }
-              in.cell = &cell;
+              fe_support_values.reinit (cell, face_idx);
 
-              this->get_material_model().evaluate(in, out);
+              // Evaluate the material model on the cell face.
+              MaterialModel::MaterialModelInputs<dim> in_support(fe_support_values, &cell, this->introspection(), this->get_solution());
+              MaterialModel::MaterialModelOutputs<dim> out_support(fe_support_values.n_quadrature_points, this->n_compositional_fields());
+              this->get_material_model().evaluate(in_support, out_support);
 
+              fe_support_values[this->introspection().extractors.velocities].get_function_values( topo_vector, stress_support_values );
+              cell->face(face_idx)->get_dof_indices (face_dof_indices);
+              for ( unsigned int i = 0; i < face_dof_indices.size(); ++i)
+                {
+                  // Given the face dof, we get the component and overall cell dof index.
+                  const std::pair<unsigned int, unsigned int> component_index = this->get_fe().face_system_to_component_index(i);
+                  const unsigned int component = component_index.first;
+                  const unsigned int support_index = component_index.second;
+                  // Dynamic topography is stored in the temperature component.
```

Comment:
Temperature is a convenient scalar finite element space to store the (scalar) dynamic topography.